src/txn/txn_recover.c

```diff
@@ -522,6 +532,15 @@ __wt_txn_recover(WT_SESSION_IMPL *session)
 done:	FLD_SET(conn->log_flags, WT_CONN_LOG_RECOVER_DONE);
 err:	WT_TRET(__recovery_free(&r));
 	__wt_free(session, config);
+
+	/*
+	 * Destroy the eviction threads that were started in support of
+	 * recovery.  They will be restarted once the lookaside table is
+	 * created.
+	 */
+	if (eviction_started)
+		WT_TRET(__wt_evict_destroy(session));
+
```

Comment:
We'd have to split recovery into pieces to do what you're suggesting -- we don't want the overhead of starting and stopping eviction every time we start, only if a real recovery pass is required.  At the moment, we only determine that in the middle of `__wt_txn_recover`.

And I disagree about `__wt_connection_workers` -- putting the start and stop in one place *avoids* the kind of ordering issues you are worried about.

To be clear, I don't love this change -- it made me sad that something like this is required.  That said, I started with the kind of solution you're describing and I don't think it's better than this.